src/ucp/core/ucp_context.c

```diff
@@ -1774,9 +1795,23 @@ static void ucp_fill_resources_reg_md_map_update(ucp_context_h context)
     ucs_memory_type_t mem_type;
     ucp_md_index_t md_index;
 
-    /* If we have a dmabuf provider for a memory type, it means we can register
-     * memory of this type with any md that supports dmabuf registration. */
     ucs_memory_type_for_each(mem_type) {
+       /* Fallback: If non-blocking registration was requested for this memory
+        * type but no MD actually supports it, treat it as if the request was
+        * not set (i.e., allow blocking-capable MDs as well). */
+        if (context->config.ext.reg_nb_mem_types & UCS_BIT(mem_type)) {
+            if (!(context->reg_nb_supported_mem_types & UCS_BIT(mem_type))) {
+                ucs_assert(context->reg_md_map[mem_type] == 0);
+
+                for (md_index = 0; md_index < context->num_mds; ++md_index) {
+                    ucp_context_update_md_maps(context, mem_type, md_index);
+                }
+            }
+        }
```

Comment:
1.  maybe we should move add mp_map flag updates to this function?
at this point we can already check which MD support nb_reg for this mem type and act accordingly. And then no need to save reg_nb_supported_mem_types on the context.
2. maybe add config param to control whether we allow fallback from nonblocking to blocking registration? default=yes